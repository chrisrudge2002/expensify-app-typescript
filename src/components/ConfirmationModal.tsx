import * as React from "react";
import { connect } from "react-redux";
import { AnyAction, Dispatch } from "redux";
import * as Modal from "react-modal";
import * as expenseActions from "../actions/expenses";

interface IConfirmationModalProps {
    description: string;
    id: string | undefined;
    isOpen: boolean;
    history: any;
    onCancelRemoval: () => void;
    startRemoveExpense: (id: string) => AnyAction;
}

export class ConfirmationModal extends React.Component<IConfirmationModalProps> {
    public render() {
        return (
            <Modal
                isOpen={!!this.props.isOpen}
                onRequestClose={this.props.onCancelRemoval}
                contentLabel="Confirm Removal"
                closeTimeoutMS={200}
                className="modal"
            >
            <h3 className="modal__title">Remove Expense</h3>
            <p className="modal__body">{`Are you sure you want to remove the expense '${this.props.description}'?`}</p>
            <button className="button button--confirm-yes" onClick={this.onConfirmRemoval}>Yes</button>
            <button className="button button--confirm-no" onClick={this.props.onCancelRemoval}>No</button>
          </Modal>
        );
    }
    private onConfirmRemoval = () => {
        if (this.props.id && this.props.startRemoveExpense) {
            this.props.startRemoveExpense(this.props.id);
            this.props.history.push("/");
        }
    }
}

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>) => ({
    startRemoveExpense: (id: string) => (dispatch(expenseActions.startRemoveExpense(id)) as AnyAction),
});

export default connect(undefined, mapDispatchToProps)(ConfirmationModal);