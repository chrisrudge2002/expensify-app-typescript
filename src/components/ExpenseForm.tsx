import * as moment from "moment";
import * as React from "react";
import { SingleDatePicker } from "react-dates";
import IExpense from "../interfaces/IExpense";

interface IExpenseFormProps extends React.Props<ExpenseForm> {
    expense?: IExpense | undefined;
    onFormSubmit?: (expense: IExpense) => void;
}

interface IExpenseFormState {
    amount: string;
    calendarFocused: boolean;
    createdAt: moment.Moment;
    description: string;
    error: string;
    note: string;
}

class ExpenseForm extends React.Component<IExpenseFormProps, IExpenseFormState> {
    constructor(props: IExpenseFormProps) {
        super(props);

        this.state = {
            amount: props.expense ? (props.expense.amount / 100).toString() : "",
            calendarFocused: false,
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            description: props.expense ? props.expense.description : "",
            error: "",
            note: props.expense ? props.expense.note : "",
        };
    }
    public render() {
        return (
            <form className="form" onSubmit={this.onFormSubmission}>
                { this.state.error && <p className="form__error">{this.state.error}</p> }
                <input
                    className="text-input"
                    type="text"
                    placeholder="Description"
                    autoFocus={true}
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    className="text-input"
                    type="text"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onCalendarFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    displayFormat="DD MMM YYYY"
                    id="d"
                />
                <textarea
                    className="textarea"
                    placeholder="Add a note for your expense (optional)"
                    value={this.state.note}
                    onChange={this.onNoteChange}>
                </textarea>
                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
        );
    }
    private onAmountChange = (e: React.FormEvent<HTMLInputElement>) => {
        const amount = e.currentTarget.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    }
    private onCalendarFocusChange = ({ focused }: { focused: boolean | null }) => {
        this.setState(() => ({ calendarFocused: focused ? focused : false }));
    }
    private onDateChange = (createdAt: moment.Moment | null) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    }
    private onDescriptionChange = (e: React.FormEvent<HTMLInputElement>) => {
        const description = e.currentTarget.value;
        this.setState(() => ({ description }));
    }
    private onNoteChange = (e: React.FormEvent<HTMLTextAreaElement>) => {
        const note = e.currentTarget.value;
        this.setState(() => ({ note }));
    }
    private onFormSubmission = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: "Please provide description and amount."}));
        } else {
            this.setState(() => ({ error: "" }));
            if (this.props.onFormSubmit) {
                this.props.onFormSubmit({
                    amount: parseFloat(this.state.amount) * 100,
                    createdAt: this.state.createdAt.valueOf(),
                    description: this.state.description,
                    note: this.state.note,
                });
            }
        }
    }
}

export default ExpenseForm;
