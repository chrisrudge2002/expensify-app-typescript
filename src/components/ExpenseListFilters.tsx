import { Moment } from "moment";
import * as React from "react";
import { DateRangePicker } from "react-dates";
import { connect } from "react-redux";
import { AnyAction, Dispatch } from "redux";
import * as filterActions from "../actions/filters";
import IAppState from "../interfaces/IAppState";
import IFilterState from "../interfaces/IFilterState";

interface IExpenseListFiltersProps extends React.Props<ExpenseListFilters> {
    filters: IFilterState;
    setEndDate: (endDate: Moment | null) => AnyAction;
    setStartDate: (startDate: Moment | null) => AnyAction;
    setTextFilter: (text: string) => AnyAction;
    sortByAmount: () => AnyAction;
    sortByDate: () => AnyAction;
}

interface IExpenseListFiltersState {
    calendarFocused: "startDate" | "endDate" | null;
}

export class ExpenseListFilters extends React.Component<IExpenseListFiltersProps, IExpenseListFiltersState> {
    public state = {
        calendarFocused: null,
    };
    public render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            className="text-input"
                            type="text"
                            value={this.props.filters.text}
                            onChange={this.onTextChange} placeholder="Search expenses"
                        />
                    </div>
                    <div className="input-group__item">
                        <select
                            className="select"
                            value={this.props.filters.sortBy}
                            onChange={this.onSortChange}>
                                <option value="date">Date</option>
                                <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            startDateId="startDateId"
                            endDate={this.props.filters.endDate}
                            endDateId="endDateId"
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onCalendarFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                            displayFormat="DD MMM YYYY"
                            showClearDates={true}
                        />
                    </div>
                </div>
            </div>
        );
    }
    private onCalendarFocusChange = (calendarFocused: "startDate" | "endDate" | null) => {
        this.setState(() => ({ calendarFocused }));
    }
    private onDatesChange = ({ startDate, endDate }: { startDate: Moment | null, endDate: Moment | null }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }
    private onSortChange = (e: React.FormEvent<HTMLSelectElement>) => {
        if (e.currentTarget.value === "date") {
            this.props.sortByDate();
        } else if (e.currentTarget.value === "amount") {
            this.props.sortByAmount();
        }
    }
    private onTextChange = (e: React.FormEvent<HTMLInputElement>) => {
        this.props.setTextFilter(e.currentTarget.value);
    }
}

const mapStateToProps = (state: IAppState) => ({
    filters: state.filters,
});

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>) => ({
    setEndDate: (endDate: Moment | null) => (dispatch(filterActions.setEndDate(endDate)) as AnyAction),
    setStartDate: (startDate: Moment | null) => (dispatch(filterActions.setStartDate(startDate)) as AnyAction),
    setTextFilter: (text: string) => (dispatch(filterActions.setTextFilter(text)) as AnyAction),
    sortByAmount: () => dispatch(filterActions.sortByAmount()),
    sortByDate: () => dispatch(filterActions.sortByDate()),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
