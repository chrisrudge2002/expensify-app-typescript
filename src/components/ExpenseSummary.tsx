import * as numeral from "numeral";
import * as React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import IAppState from "../interfaces/IAppState";
import selectExpenses from "../selectors/expenses";
import selectExpensesTotal from "../selectors/expenses-total";

interface IExpenseSummaryProps {
    expenseCount: number;
    totalAmount: number;
}

export const ExpenseSummary: React.SFC<IExpenseSummaryProps> = ({ expenseCount, totalAmount }) => {
    const expenseWord = expenseCount === 1 ? "expense" : "expenses";
    const formattedTotal = numeral(totalAmount / 100).format("$0,0.00");
    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-hesder__title">
                    Viewing <span>{expenseCount}</span> {expenseWord} totalling <span>{formattedTotal}</span>
                </h1>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state: IAppState) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        totalAmount: selectExpensesTotal(visibleExpenses),
    };
};

export default connect(mapStateToProps)(ExpenseSummary);
