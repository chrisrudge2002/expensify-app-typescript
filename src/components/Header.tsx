import * as React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { AnyAction, Dispatch } from "redux";
import * as authActions from "../actions/auth";

interface IHeaderProps {
    startLogout: () => AnyAction;
}

export const Header: React.SFC<IHeaderProps> = (props) => (
    <header className="header">
        <div className="content-container">
            <div className="header__content">
                <Link className="header__title" to="/dashboard">
                    <h1>Expensify</h1>
                </Link>
                <button className="button button--link" onClick={props.startLogout}>Logout</button>
            </div>
        </div>
    </header>
);

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>) => ({
    startLogout: () => dispatch(authActions.startLogout()),
});

export default connect(undefined, mapDispatchToProps)(Header);
