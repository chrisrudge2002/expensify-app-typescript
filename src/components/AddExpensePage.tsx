import * as React from "react";
import { connect } from "react-redux";
import { RouterProps } from "react-router";
import { AnyAction, Dispatch } from "redux";
import * as expenseActions from "../actions/expenses";
import IExpense from "../interfaces/IExpense";
import ExpenseForm from "./ExpenseForm";

interface IAddExpensePageProps extends RouterProps {
    startAddExpense: (expense: IExpense) => AnyAction;
}

export class AddExpensePage extends React.Component<IAddExpensePageProps> {
    public render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm onFormSubmit={this.onSubmit} />
                </div>
            </div>
        );
    }
    private onSubmit = (expense: IExpense) => {
        this.props.startAddExpense(expense);
        this.props.history.push("/");
    }
}

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>) => ({
    startAddExpense: (expense: IExpense) => (dispatch(expenseActions.startAddExpense(expense)) as AnyAction),
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
