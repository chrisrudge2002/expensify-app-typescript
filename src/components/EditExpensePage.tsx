import * as React from "react";
import { connect } from "react-redux";
import { RouteComponentProps } from "react-router";
import { AnyAction, Dispatch } from "redux";
import * as expenseActions from "../actions/expenses";
import IAppState from "../interfaces/IAppState";
import IExpense from "../interfaces/IExpense";
import ExpenseForm from "./ExpenseForm";
import ConfirmationModal from "./ConfirmationModal";

interface IEditExpensePageProps extends RouteComponentProps<any> {
    expense: IExpense;
    startEditExpense: (id: string, updates: {} ) => AnyAction;
}

interface IEditExpensePageState {
    confirmModalIsOpen: boolean;
}

export class EditExpensePage extends React.Component<IEditExpensePageProps, IEditExpensePageState> {
    constructor(props: IEditExpensePageProps) {
        super(props);

        this.state = {
            confirmModalIsOpen: false,
        };
    }
    public render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm expense={this.props.expense} onFormSubmit={this.onSubmit} />
                    <button
                        className="button button--secondary"
                        onClick={this.onRemove}>
                            Remove Expense
                    </button>
                </div>
                <ConfirmationModal 
                    id={this.props.expense.id}
                    description={this.props.expense.description}
                    isOpen={this.state.confirmModalIsOpen}
                    onCancelRemoval={this.onCancelRemoval}
                    history={this.props.history} />
            </div>
        );
    }
    private onCancelRemoval = () => {
        this.setState(() => ({ confirmModalIsOpen: false }));
    }
    private onSubmit = (expense: IExpense) => {
        if (this.props.expense.id) {
            this.props.startEditExpense(this.props.expense.id, expense);
            this.props.history.push("/");
        }
    }
    private onRemove = () => {
        if (this.props.expense.id) {
            this.setState(() => ({ confirmModalIsOpen: true }));
        }
    }
}

const mapStateToProps = (state: IAppState, props: IEditExpensePageProps) => ({
    expense: state.expenses.find((e: IExpense) => e.id === props.match.params.id),
});

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>) => ({
    startEditExpense: (id: string, expense: IExpense) =>
        (dispatch(expenseActions.startEditExpense(id, expense)) as AnyAction),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
