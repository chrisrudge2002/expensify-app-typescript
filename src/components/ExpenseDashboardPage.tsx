import * as React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpenseSummary from "./ExpenseSummary";

const ExpenseDashboardPage: React.SFC<any> = () => (
    <div>
        <ExpenseSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;
