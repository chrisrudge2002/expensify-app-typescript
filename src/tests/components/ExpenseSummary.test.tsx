import { shallow } from "enzyme";
import * as React from "react";
import { ExpenseSummary } from "../../components/ExpenseSummary";

test("should render ExpenseSummary correctly with no expenses", () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={0} totalAmount={0} />);
    expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseSummary correctly with one expense", () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={1} totalAmount={4500} />);
    expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseSummary correctly with multiple expenses", () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={2} totalAmount={1234500} />);
    expect(wrapper).toMatchSnapshot();
});
