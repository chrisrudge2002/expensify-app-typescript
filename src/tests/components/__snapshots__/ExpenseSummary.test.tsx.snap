// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary correctly with multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-hesder__title"
    >
      Viewing 
      <span>
        2
      </span>
       
      expenses
       totalling 
      <span>
        $12,345.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseSummary correctly with no expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-hesder__title"
    >
      Viewing 
      <span>
        0
      </span>
       
      expenses
       totalling 
      <span>
        $0.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseSummary correctly with one expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-hesder__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $45.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
